packages/governance/src/governor/governor.cairo

```diff
@@ -0,0 +1,959 @@
+                Errors::RESTRICTED_PROPOSER
+            );
+
+            // Check proposal threshold
+            let vote_threshold = self._proposal_threshold();
+            if vote_threshold > 0 {
+                let votes = self
+                    ._get_votes(proposer, self.clock() - 1, Immutable::DEFAULT_PARAMS());
+                assert(votes >= vote_threshold, Errors::INSUFFICIENT_PROPOSER_VOTES);
+            }
+
+            self._propose(calls, @description, proposer)
+        }
+
+        /// Queues a proposal. Some governors require this step to be performed before execution can
+        /// happen. If queuing is not necessary, this function may revert.
+        /// Queuing a proposal requires the quorum to be reached, the vote to be successful, and the
+        /// deadline to be reached.
+        ///
+        /// Returns the id of the proposal.
+        ///
+        /// Requirements:
+        ///
+        /// - The proposal must be in the `Succeeded` state.
+        /// - The queue operation must return a non-zero ETA.
+        ///
+        /// Emits a `ProposalQueued` event.
+        fn queue(
+            ref self: ComponentState<TContractState>, calls: Span<Call>, description_hash: felt252
+        ) -> felt252 {
+            let proposal_id = self._hash_proposal(calls, description_hash);
+            self.validate_state(proposal_id, array![ProposalState::Succeeded].span());
+
+            let eta_seconds = self.queue_operations(proposal_id, calls, description_hash);
+            assert(eta_seconds > 0, Errors::QUEUE_NOT_IMPLEMENTED);
+
+            let mut proposal = self.Governor_proposals.read(proposal_id);
+            proposal.eta_seconds = eta_seconds;
+            self.Governor_proposals.write(proposal_id, proposal);
+
+            self.emit(ProposalQueued { proposal_id, eta_seconds });
+
+            proposal_id
+        }
+
+        /// Executes a successful proposal. This requires the quorum to be reached, the vote to be
+        /// successful, and the deadline to be reached. Depending on the governor it might also be
+        /// required that the proposal was queued and that some delay passed.
+        ///
+        /// NOTE: Some modules can modify the requirements for execution, for example by adding an
+        /// additional timelock (See `timelock_controller`).
+        ///
+        /// Returns the id of the proposal.
+        ///
+        /// Requirements:
+        ///
+        /// - The proposal must be in the `Succeeded` or `Queued` state.
+        ///
+        /// Emits a `ProposalExecuted` event.
+        fn execute(
+            ref self: ComponentState<TContractState>, calls: Span<Call>, description_hash: felt252
+        ) -> felt252 {
+            let proposal_id = self._hash_proposal(calls, description_hash);
+            self
+                .validate_state(
+                    proposal_id, array![ProposalState::Succeeded, ProposalState::Queued].span()
+                );
```

Comment:
It seems to be overkill for checking the state: we have to create an array and iterate through it. Maybe a more direct approach would be better 

```suggestion
  match self.state(proposal_id) {
    ProposalState::Succeeded => (),
    ProposalState::Queued => (),
    _ => panic_with_felt252(Errors:: UNEXPECTED_PROPOSAL_STATE),
```

Reply:
This is probably more efficient, but the validate_state helper can be used by users as well. We can refactor it into bitmap operations to avoid looping or matching, which should be even more efficient, and also usable by users. Let's keep the thread opened until the pending stuff is finished and I will work on that then.